calour/ratio_experiment.py

```diff
@@ -0,0 +1,254 @@
+        The data values for each feature are the ratio of mean data values in samples of value1 in group_field / data values in samples of value2 in group_field
+        '''
+        new_field_val = '%s / ' % value1
+        exp.sparse = False
+        value1 = _to_list(value1)
+        if value2 is None:
+            new_field_val += 'Other'
+            value2 = list(set(exp.sample_metadata[group_field].unique()).difference(set(value1)))
+        else:
+            new_field_val += '%s' % value2
+            value2 = _to_list(value2)
+
+        ratio_mat = np.zeros([len(exp.sample_metadata[common_field].unique()), exp.shape[1]])
+        # new_sample_metadata = pd.DataFrame()
+        new_sample_metadata = pd.DataFrame(columns=exp.sample_metadata.columns)
+
+        # used to count the number of samples that have values in both groups
+        # we then keep only these columns in the ratio_mat
+        found_indices = []
+        for idx, cexp in enumerate(exp.iterate(common_field, axis=0)):
+            # cfield = cexp.sample_metadata[common_field].iloc[0]
+            group1 = cexp.filter_samples(group_field, value1)
+            group2 = cexp.filter_samples(group_field, value2)
+            if group1.shape[0] == 0 or group2.shape[0] == 0:
+                continue
+            # replace all <= threshold values if needed
+            if threshold is not None:
+                group1.data[group1.data <= threshold] = threshold
+                group2.data[group2.data <= threshold] = threshold
+
+            mean_g1 = np.mean(group1.data, axis=0)
+            mean_g2 = np.mean(group2.data, axis=0)
+            ratios = np.log2(mean_g1 / mean_g2)
+            # Take care of features where both nominator and denominator are below the min_threshold. We should get np.nan
+            if threshold is not None:
+                ratios[np.logical_and(mean_g1 == threshold, mean_g2 == threshold)] = np.nan
+            ratio_mat[idx, :] = ratios
+            cmetadata = cexp.sample_metadata.iloc[0]
+            for ccol in group1.sample_metadata.columns:
+                u1 = group1.sample_metadata[ccol].unique()
+                u2 = group2.sample_metadata[ccol].unique()
+                if len(u1) == 1 and u1 == u2:
+                    cmetadata.at[ccol] = u1[0]
+                else:
+                    cmetadata.at[ccol] = 'NA (multiple values)'
+            cmetadata.at[group_field] = new_field_val
+            new_sample_metadata = new_sample_metadata.append(cmetadata)
+            found_indices.append(idx)
+
+        # keep only samples that were actually added to the ratio_mat
+        ratio_mat = ratio_mat[found_indices, :]
+        ratio_exp = RatioExperiment(data=ratio_mat, sample_metadata=new_sample_metadata, feature_metadata=exp.feature_metadata,
+                                    sparse=False, databases=exp.databases, description=exp.description, info=exp.info)
+        return ratio_exp
```

Comment:
this implementation feels complicated. can we filter before treatment samples and after treatment samples? and sort them the same by subject, and finally divide.

Reply:
don't see how to do it simpler. can you give more details?